src/plugins/undoRedo/undoRedo.js

```diff
@@ -45,167 +52,199 @@ function UndoRedo(instance) {
-    }
 
-    const columnsMap = instance.columnIndexMapper.getIndexesSequence();
-    const rowsMap = instance.rowIndexMapper.getIndexesSequence();
-    const action = new UndoRedo.RemoveColumnAction(
-      columnIndex, indexes, removedData, headers, columnsMap, rowsMap, instance.getSettings().fixedColumnsLeft);
+      if (Array.isArray(instance.getSettings().colHeaders)) {
+        rangeEach(amount - 1, (i) => {
+          headers.push(instance.getSettings().colHeaders[instance.toPhysicalColumn(columnIndex + i)] || null);
+        });
+      }
+
+      const columnsMap = instance.columnIndexMapper.getIndexesSequence();
+      const rowsMap = instance.rowIndexMapper.getIndexesSequence();
+
+      return new UndoRedo.RemoveColumnAction(
+        columnIndex, indexes, removedData, headers, columnsMap, rowsMap, instance.getSettings().fixedColumnsLeft);
+    };
 
-    plugin.done(action);
+    plugin.done(wrappedAction, source);
   });
 
   instance.addHook('beforeCellAlignment', (stateBefore, range, type, alignment) => {
-    const action = new UndoRedo.CellAlignmentAction(stateBefore, range, type, alignment);
-
-    plugin.done(action);
+    plugin.done(() => new UndoRedo.CellAlignmentAction(stateBefore, range, type, alignment));
   });
 
   instance.addHook('beforeFilter', (conditionsStack) => {
-    plugin.done(new UndoRedo.FiltersAction(conditionsStack));
+    plugin.done(() => new UndoRedo.FiltersAction(conditionsStack));
   });
 
   instance.addHook('beforeRowMove', (rows, finalIndex) => {
     if (rows === false) {
       return;
     }
 
-    plugin.done(new UndoRedo.RowMoveAction(rows, finalIndex));
+    plugin.done(() => new UndoRedo.RowMoveAction(rows, finalIndex));
   });
 
   instance.addHook('beforeMergeCells', (cellRange, auto) => {
     if (auto) {
       return;
     }
 
-    plugin.done(new UndoRedo.MergeCellsAction(instance, cellRange));
+    plugin.done(() => new UndoRedo.MergeCellsAction(instance, cellRange));
   });
 
   instance.addHook('afterUnmergeCells', (cellRange, auto) => {
     if (auto) {
       return;
     }
 
-    plugin.done(new UndoRedo.UnmergeCellsAction(instance, cellRange));
+    plugin.done(() => new UndoRedo.UnmergeCellsAction(instance, cellRange));
   });
-
 }
 
 /**
- * @param {object} action The action desciptor.
+ * Stash information about performed actions.
+ *
+ * @function done
+ * @memberof UndoRedo#
+ * @fires Hooks#beforeUndoStackChange
+ * @fires Hooks#afterUndoStackChange
+ * @fires Hooks#beforeRedoStackChange
+ * @fires Hooks#afterRedoStackChange
+ * @param {Function} wrappedAction The action descriptor wrapped in a closure.
+ * @param {string} [source] Source of the action. It is defined just for more general actions (not related to plugins).
  */
-UndoRedo.prototype.done = function(action) {
-  if (!this.ignoreNewActions) {
-    this.doneActions.push(action);
-    this.undoneActions.length = 0;
+UndoRedo.prototype.done = function(wrappedAction, source) {
+  if (this.ignoreNewActions) {
```

Comment:
@wszymanski @budnix I think we have this place in mind in context of https://github.com/handsontable/handsontable/pull/7827#issuecomment-824217906
```suggestion
  if (this.ignoreNewActions && isInternalUndoAction) {
```

Reply:
Done something similar: https://github.com/handsontable/handsontable/pull/7827/files#diff-d5e9d9bed12124bb818798c7c92975beea688e6867ad41c1c893a9ea946110d7R184